### ISSUE TITLE
Kevo API Issue - Error: Invalid response code 500 when trying to log in or get/set status

### ISSUE THREAD

**Issue by thatguychad** [nfarina/homebridge-kevo#11]
I've been seeing this error more frequently since March 22nd. Has Kevo changed their login method, breaking the plugin? My locks are showing "no response" almost constantly now.

**Comment by Tiger519** [comment]
I'm seeing the same thing. It started happening to me back in July 2017, but only for about 3 days. I believe I started experiencing these issues again about the same time you did. Because it works _sometimes_, that makes me think it's not a login issue per se. In fact, if you go to www.mykevo.com and just refresh the page a bunch of times, you'll eventually get and internal server error, though not nearly as often as I do in Homebridge. I'm not sure why it happens so often in Homebridge compared to the website, and I definitely don't understand why the Kevo app itself never seems to have issues.

That said, I suspect it has to do with the frequency of the queries. For some reason, most of the time I see double entries in Homebridge like so:

[4/9/2018, 10:13:35 PM] [Bedroom Door] Getting current state...
[4/9/2018, 10:13:35 PM] [Bedroom Door] Getting current state...
[4/9/2018, 10:13:35 PM] [Bedroom Door] Already logged in.
[4/9/2018, 10:13:35 PM] [Bedroom Door] Already logged in.
[4/9/2018, 10:13:36 PM] [Bedroom Door] Lock status is Locked
[4/9/2018, 10:13:36 PM] [Bedroom Door] Lock status is Locked

Now this time it worked, but often times one of the requests fails, which causes the lock to report not responding (it can fail at login and also at getting/setting lock status). I also have 4 locks, so that amplifies it on my end. But even when I'm querying just one lock status, I still get failures rather frequently (but not as frequently as when I'm pulling all 4 from my Home tab).

At the end of the day, I think this is wholly on Kevo's side. I'm a little concerned with the "double" status in Homebridge, but it was doing the same thing when everything was working perfectly, so I really don't think that's it.  It's also not queries over time, because I had a gap of 12 hours between queries and still had failures on every lock. So either their website is just having random failures which the phone app somehow "filters out", or they have introduced some method to determine if a call to those JSONs is from their app or another source (i.e. Homebridge). Either way I fear there isn't anything we cann do about it, but I would love to be proven wrong!

**Comment by Tiger519** [comment]
So I kind of hacked together a solution to the issue, though not being a "real" programmer it's not very elegant and I'm sure there's a better way to do it. I used https://github.com/FGRibreau/node-request-retry as a drop-in replacement for request (just added it on my local). At first it wasn't loading the plugin at all, it was almost as if the extra options in the request statement (maxAttempts and retryDelay) were messing with the service definitions. So I just went into the requestretry code and changed the defaults from 5 attempts/5s delay to 15 attempts/0.05s delay. There's still a good bit of delay when you query multiple locks (was taking about 10 seconds to query all 4 of my locks at once), but the HTTP 500 errors are gone! So I think I'm probably ok with this for now.

Like I said, I'm not a "real" programmer and clearly this (modifying code for a required module because I couldn't get it to work properly!) is not a scalable method I could deploy out to people. But if anyone other there has the expertise to take this method and do it right, that would be awesome.

I don't consider this to be fixed, so I'd leave this and my related issue #9 open, though as this plugin hasn't been updated in 2 years I'm not really expecting anyone to pick this up.

**Comment by nfarina** [comment]
Hey all - sorry I'm not much help here - I haven't owned Kevo for some time so I wouldn't be able to test any changes to the repo. If someone wants to volunteer taking over as a maintainer that would probably be the best way to go.

**Comment by Tiger519** [comment]
Appreciate the response! I saw in a previous comment you didn't have your lock anymore, so I totally understand. I'm grateful you put this together in the first place! I don't have enough knowledge about js in general or even github to take this on, but if we can find someone else willing to maintain I'd be happy to help out on code snippets.

In the meantime, hopefully my hacky solution can help some people, and maybe someone will be inspired to put together a more permanent solution :-)

**Comment by thatguychad** [comment]
@Tiger519 Did you just do a normal npm install on requestretry? Can you post more specific instructions on what you did, please?

**Comment by Tiger519** [comment]
@thatguychad Yes, I did npm install. I can't remember where I did the command from but I did get it into the homebridge-kevo/node_modules folder. I did it again as npm install --save requestretry to add to package.js, though since I did it all on local I don't think that was necessary. Then I replaced "request" with "requestretry" in the variable definition. I also moved the urls into the options array (e.g. {url:url, form:form} instead of url, {form:form}) though I don't know if that was necessary. When I added maxAttempts and retryDelay to the options, the plugin failed to load. I started to troubleshoot but when it occurred to me I could just modify the defaults within requestretry, I just went into that module's index.js and changed the defaults to 15/50. Later I went back and changed it to 100/50 (thought it was failing after 15 tried, though turned out it was a temporary issue with the Kevo servers yesterday.

I'm pretty new to nodejs and npm, but I think by adding the new dependency to the package, it will be installed with the plugin. So really the remaining issue to work out is adding the options to the request options without causing the plugin to fail loading. But I'm guessing you would just like to get this working too, so hopefully the instructions above can get you going in the right direction.

I tried the original defaults, but it was too slow considering there's a login request AND a status request, and it seemed each one was running 5-8 times (and I have 4 locks, though I think each lock status request is handled in parallel). Also, it seems that usually each query is run twice, and Homekit doesn't resolve the status until both responses are received. No idea why it queries twice, it doesn't happen every time but I've seen the same thing in logs from other users (not related to this change). I've always assumed it's due to the async nature of nodejs causing Homekit to try the query twice, since I can't see anything in the code that would make it query twice.

**Comment by pmeisel123** [comment]
I have written a quick little hack to handle this.  I am a double noob here, for writing code homebridge and to git/github.  So, my solution is not elegant, and I can not figure out how to make a change request from my local copy to here.  

Here is a copy of my local change:
https://github.com/pmeisel123/homebridge-kevo/commit/412d7bc982c7f28e7a07901935028a65ff913b66

Basically I built a real simple wrapper for request, that says make the request check the response status if it is a 500 try again (up to 5 times).  

Note: I am fine with someone else taking my code, cleaning it up, and submitting it.

**Comment by ItsDLite** [comment]
Could I get some help with this or is it dead? Loaded plugin: homebridge-kevo and installed the two dependencies cheerio and request. Updated homebridge Config.json - homebridge and iOS Home works on my Smartthings applications. But I don't see my lock and can't use it (*Yes I have Kevo Plus) Help Please ~ David

**Comment by Tiger519** [comment]
I also tried my hand at updating this, https://github.com/Tiger519/homebridge-kevo. I must have messed up something else with my initial attempt that caused the plugin not to load, everything seems fine now with the custom parameters.

However, now that I get this figured out, the Kevo website seems to be working more reliably again :-). So I'm not sure if this adds additional overhead that is unnecessary or if it's better to build something in for the next time the website inevitably starts returning HTTP 500s again...

One other note, because I'm super new at this, and also because I don't want to mess up my existing config, I didn't actually try to clone my new repository. I made the changes directly in the index.js of the original plug-in to test, then replicated in my forked repository. That said, this was pretty straightforward, but someone should definitely try to do some independent testing before we consider merging this change. Probably won't get much attention until the website starts returning bad responses again!

Edit: My fork no longer has Requestretry in it. I started from a new master to fix an issue with multiple locks. I plant to create another branch within my repo with these updates, but since the Kevo API has calmed down I've gone back to using the base plugin.

**Comment by Tiger519** [comment]
I have begun experiencing this issue again today. I still have a version of the code replacing request with requestretry on my local, I'm going to sync it up with the latest master and test it for a few days to see if it is reliable enough to push. I have concerns that this may impact performance, especially for those with multiple locks like me, as you can be making quite a lot of calls in a short period of time. Even if that is so, it's probably better than suffering through days or weeks of HTTP 500 errors for something critical like door locks.

**Comment by 4G-Gregg** [comment]
Hey @Tiger519 I'm experiencing the same issue and also looked to adding requestretry. Did you have any luck with this method?

**Comment by Tiger519** [comment]
Yes, I did. It's still running on my local. Let me see about pushing it to the master. May take me a week or so to get to it, however.

**Comment by 4G-Gregg** [comment]
Would you mind throwing it in a branch so that I could test it as well?

**Comment by Tiger519** [comment]
@4G-Gregg I finally got around to doing this. I have updated my fork at tiger519/homebridge-kevo with the code to use requestretry. I do not have the time to test this as a full install via npm. I initially just edited the changes into the index.js file on my local install of the master. In my fork I also added requestretry as a dependency in the package.json, I've never needed to update that before but I'm assuming this will ensure requestretry makes it over as part of the npm install (if I ened to do soemthing different here let me know). I'm pretty sure the code all made it over ok, but it was done manually so if there is any issue running it please let me know and I should be able to fix quickly this week. I'll depend on your testing before we merge this to the master.
